src/realm/string_interner.hpp

```diff
@@ -37,19 +37,40 @@ namespace realm {
 class StringCompressor;
 
 struct CachedString {
-    uint8_t m_weight = 0;
+    std::atomic<uint8_t> m_weight = 0;
     std::unique_ptr<std::string> m_decompressed;
+    CachedString() {}
+    CachedString(CachedString&& other)
+    {
+        m_decompressed = std::move(other.m_decompressed);
+        m_weight.store(other.m_weight.load(std::memory_order_relaxed), std::memory_order_relaxed);
+    }
+    CachedString(uint8_t init_weight, std::unique_ptr<std::string>&& ptr)
+    {
+        m_decompressed = std::move(ptr);
+        m_weight.store(init_weight, std::memory_order_relaxed);
+    }
 };
 
 class StringInterner {
 public:
-    // To be used exclusively from Table
+    // Use of the StringInterner must honour the restrictions on concurrency given
+    // below. Currently this is ensured by only using concurrent access on frozen
+    // objects.
+    //
+    // Limitations wrt concurrency:
+    //
+    // To be used exclusively from Table and in a non-concurrent setting:
```

Comment:
Would it make sense to have a counter that is increased/decreased when calling the non-threadsafe operation. Then we could have assertions requiring that the counter is 1 inside those functions and 0 in the threadsafe functions below.

Reply:
You mean for asserts?

Reply:
for such a counter to be correctly shared among threads (and thus be useful for asserting) it would have to be an atomic which carries some overhead. I'm not convinced it is worth it?

Reply:
I could add it in debug mode only?